replicator/run.py

```diff
@@ -0,0 +1,161 @@
+        #  falls behind at the same rate if there is high load. Otherwise, the order is random, and certain domains
+        #  may end up having a disadvantage in AXFR scheduling. (Not sure if this is the case.)
+        #  On the other hand, if several changes are done quickly on the same domain (e.g. delete/recreate domain),
+        #  fetching the new catalog here allows skipping useless intermediate states. Also, in case processing somehow
+        #  is stuck because the catalog is invalid, always fetching a new catalog improves chances of recovery.
+        #  Also, if this is not always done, then long-running AXFRs may hold up new changes to the catalog, as we
+        #  would be waiting for large zone's serial to update until we retrieve a new catalog.
+        catalog = pdns_request('get', path=f'/zones/{catalog_domain_id}').json()
+        local_catalog_serial = catalog['serial']
+
+        # Update the catalog in the background if necessary
+        remote_catalog_serial = query_serial(catalog_domain)
+        if local_catalog_serial != remote_catalog_serial:
+            print(f'Queueing catalog AXFR from {local_catalog_serial} to {remote_catalog_serial} ...')
+            pdns_request('put', path=f'/zones/{catalog_domain_id}/axfr-retrieve')
+
+        # See if the last parsed catalog is the current one
+        if serials[catalog_domain] == local_catalog_serial:
+            print(f'Nothing to do (catalog {serials[catalog_domain]} with {len(serials)} zones up to date).')
+            sleep(1)
+            continue
+
+        # Parse current state of catalog zone and determine slave-side changes
+        catalog_map = {}
+        for rrset in catalog['rrsets']:
+            m = re.match(r'^(serial)?\.?([0-9a-f]{40}).zones.catalog.internal.$', rrset['name'])
+            if m is None:
+                continue
+
+            subname, m_unique = m.groups()
+
+            map = catalog_map.setdefault(m_unique, {})
+            if rrset['type'] == 'PTR' and subname is None:
+                map['name'] = rrset['records'][0]['content']
+            elif rrset['type'] == 'TXT' and subname == 'serial':
+                map['serial'] = int(rrset['records'][0]['content'][1:-1])
+            else:
+                raise Exception('Unexpected catalog rrset %s' % rrset)
+
+        if not all(len(map) == 2 for map in catalog_map.values()):
+            raise Exception('Catalog has a value of m_unique with inconsistent content')
+
+        local_zones = set(name for name in serials.keys() if name != catalog_domain)
+        catalog_zones = set(map['name'] for map in catalog_map.values())
+
+        deletions = local_zones - catalog_zones
+        additions = catalog_zones - local_zones
+        modifications = {map['name']: map['serial'] for map in catalog_map.values()
+                         if map['name'] in serials and map['serial'] != serials[map['name']]}
```

Comment:
Per line 118, content of `catalog_zones` and `catalog_map` contains the same names. Let say, `example.com` is in `local_zones`, but not in `catalog_zones` and `catalog_map`. Would `example.com` then end up in `deletions` and `modifications`? Can we assert that the three sets `deletions`, `additions`, and `modifications` are a partition of all known domains?

Reply:
Can the catalog zone end up in `modifications`?

Reply:
Regarding the first question: `modifications` is created by iterating over `catalog_map` which has information on domains listed in the catalog. Thus, if `example.com` is in `local_zones` but not in `catalog_zones`/`catalog_map`, it would end up in `deletions` only.

What do you mean by asserting that the three sets (not that `modifications` is actually a `dict`) are a partition of all known domains? There's also the `unchanged` group, which is not computed explicitly here.

Reply:
Regarding the second question: The catalog zone can only end up in `modifications` if it is listed itself in the catalog. However, this is not the case, as the catalog resides on `nsmaster` only (not `nslord`) and is created by iterating over all zones known on `nslord`. In other words, the catalog zone does not contain itself, and therefore it will never end up in modifications.

However, because it is not listed in the catalog, one has to take precautions to not accidentally delete the catalog on the slave. This is why the zone is explicitly exempt from the `local_zones` set (it's handled in the line where that variable is assigned).

Reply:
obsolete as the code has been drastically simplified